keras_cv/models/segmentation/segment_anything/sam_prompt_encoder.py

```diff
@@ -0,0 +1,321 @@
+    downscale it to generate "dense encodings". If no mask prompt is passed,
+    an embedding layer is used instead to generate a "no mask" embedding.
+
+    Args:
+        embed_dim (int): The number of features in the output embeddings.
+        image_embedding_size (int): The number of features in the image
+            embeddings generated by an image encoder.
+        input_image_size (tuple[int, int]): A tuple of the height and width
+            of the image being prompted.
+        mask_in_chans (int): The number of channels of the mask prompt.
+        activation (str, optional): The activation to use in the mask
+            downscaler neural net. Defaults to "gelu".
+
+    References:
+        - [Segment Anything](https://arxiv.org/abs/2304.02643)
+    """
+
+    def __init__(
+        self,
+        embed_dim,
+        image_embedding_size,
+        input_image_size,
+        mask_in_chans,
+        activation="gelu",
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.embed_dim = embed_dim
+        self.image_embedding_size = image_embedding_size
+        self.input_image_size = input_image_size
+        self.mask_in_chans = mask_in_chans
+        self.activation = activation
+
+        self.positional_embedding_layer = RandomFrequencyPositionalEmbeddings(
+            num_positional_features=self.embed_dim // 2, scale=1
+        )
+
+        self.foreground_point_embed = keras.layers.Embedding(1, embed_dim)
+        self.background_point_embed = keras.layers.Embedding(1, embed_dim)
+        self.top_left_corner_embed = keras.layers.Embedding(1, embed_dim)
+        self.bottom_right_corner_embed = keras.layers.Embedding(1, embed_dim)
+        self.not_a_point_embed = keras.layers.Embedding(1, embed_dim)
+
+        self.mask_downscaler = SerializableSequential(
+            [
+                keras.layers.Conv2D(
+                    mask_in_chans // 4, kernel_size=2, strides=2
+                ),
+                keras.layers.LayerNormalization(epsilon=1e-6),
+                keras.layers.Activation(activation),
+                keras.layers.Conv2D(mask_in_chans, kernel_size=2, strides=2),
+                keras.layers.LayerNormalization(epsilon=1e-6),
+                keras.layers.Activation(activation),
+                keras.layers.Conv2D(embed_dim, kernel_size=1),
+            ]
+        )
+        self.mask_downscaler.build(
+            [None, 4 * image_embedding_size[0], 4 * image_embedding_size[1], 1]
+        )
+        self.no_mask_embed = keras.layers.Embedding(1, embed_dim)
+
+        for layer in [
+            self.foreground_point_embed,
+            self.background_point_embed,
+            self.top_left_corner_embed,
+            self.bottom_right_corner_embed,
+            self.not_a_point_embed,
+            self.no_mask_embed,
+        ]:
+            layer.build(None)
+
+        self.built = True
+
+    def get_dense_pe(self):
```

Comment:
I think this should be a private method, the same as the positional embeddings for the previous layer - or potentially turned into a layer themselves?

Reply:
Removed it and made it a part of the call method itself to make the inputs and output well-defined for the Task model.